File: Classes/PreyStore/PreyStoreManager.m
//
//  PreyStoreManager.m
//  Prey
//
//  Created by Javier Cala Uribe on 7/7/15.
//  Copyright (c) 2015 Fork Ltd. All rights reserved.
//

#import "PreyStoreManager.h"
#import "PreyStoreProduct.h"
#import "SFHFKeychainUtils.h"

@implementation PreyStoreManager

#pragma mark Init

+ (PreyStoreManager*)instance
{
    static PreyStoreManager *instance = nil;
    static dispatch_once_t onceToken = 0;
    
    dispatch_once(&onceToken, ^{
        instance = [[PreyStoreManager alloc] init];
        instance.purchasableObjects = [NSMutableArray array];
        [instance requestProductData];
        
        [[SKPaymentQueue defaultQueue] addTransactionObserver:instance];
    });

    return instance;
}



- (void)requestProductData
{
    NSMutableSet *products = [[NSMutableSet alloc] init];
    [products addObjectsFromArray:[NSArray arrayWithObjects:kSubscription1Year, nil]];
    
    self.productsRequest = [[SKProductsRequest alloc] initWithProductIdentifiers:products];
    self.productsRequest.delegate = self;
    [self.productsRequest start];
    
    PreyLogMessage(@"PreyStore Manager", 10, @"Start SKProductsRequest");
}

- (void)buyFeature:(NSString*)featureId onComplete:(void (^)(NSString*, NSData*, NSArray*))completionBlock
       onCancelled:(void (^)(void))cancelBlock
{
    self.onTransactionCompleted = completionBlock;
    self.onTransactionCancelled = cancelBlock;
    
    [self addToQueue:featureId];
}

- (void)addToQueue:(NSString*)productId
{
    if ([SKPaymentQueue canMakePayments])
    {
        NSArray   *allIds = [self.purchasableObjects valueForKey:@"productIdentifier"];
        NSUInteger  index = [allIds indexOfObject:productId];
        
        if (index == NSNotFound) return;
        
        SKProduct *thisProduct = [self.purchasableObjects objectAtIndex:index];
        SKPayment *payment = [SKPayment paymentWithProduct:thisProduct];
        [[SKPaymentQueue defaultQueue] addPayment:payment];
    }
    else
    {
        [self showAlertWithTitle:NSLocalizedString(@"In-App Purchasing disabled", @"")
                         message:NSLocalizedString(@"Check your parental control settings and try again later", @"")];
    }
}

- (void)showAlertWithTitle:(NSString*)title message:(NSString*)message
{    
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:title
                                                    message:message
                                                   delegate:nil
                                          cancelButtonTitle:NSLocalizedString(@"Dismiss", @"")
                                          otherButtonTitles:nil];
    [alert show];
}

- (void)restorePreviousTransactionsOnComplete:(void (^)(void))completionBlock onError:(void (^)(NSError*))errorBlock
{
    self.onRestoreCompleted = completionBlock;
    self.onRestoreFailed = errorBlock;
    
    [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
}


#pragma mark Delegate SKProductsRequest

- (void)productsRequest:(SKProductsRequest *)request didReceiveResponse:(SKProductsResponse *)response
{
    PreyLogMessage(@"PreyStore Manager", 10, @"DidReceiveResponse SKProductsRequest");
    
    [self.purchasableObjects addObjectsFromArray:response.products];
    
#ifdef DEBUG
    for (int i=0;i<[self.purchasableObjects count];i++)
    {
        SKProduct *product = [self.purchasableObjects objectAtIndex:i];
        NSLog(@"Feature: %@, Cost: %f, ID: %@",[product localizedTitle], [[product price] doubleValue], [product productIdentifier]);
    }
    
    for (NSString *invalidProduct in response.invalidProductIdentifiers)
        NSLog(@"Problem in iTunes connect configuration for product: %@", invalidProduct);
#endif
    
    self.productsRequest = nil;
}

- (void)request:(SKRequest *)request didFailWithError:(NSError *)error
{
    PreyLogMessage(@"PreyStore Manager", 10, @"didFailWithError SKProductsRequest : %@",error);
    
    self.productsRequest = nil;
}


- (void)requestDidFinish:(SKRequest *)request
{
    PreyLogMessage(@"PreyStore Manager", 10, @"requestDidFinish SKProductsRequest");
}

#pragma mark Delegate SKPaymentTransactionObserver

- (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray *)transactions
{
    PreyLogMessage(@"PreyStore Manager", 10, @"updatedTransactions SKPaymentQueue");
    
    for (SKPaymentTransaction *transaction in transactions)
    {
        switch (transaction.transactionState)
        {
            case SKPaymentTransactionStatePurchased:
                
                [self completeTransaction:transaction];
                
                break;
                
            case SKPaymentTransactionStateFailed:
                
                [self failedTransaction:transaction];
                
                break;
                
            case SKPaymentTransactionStateRestored:
                
                [self restoreTransaction:transaction];
                
            default:
                break;
        }
    }
}

- (void)paymentQueue:(SKPaymentQueue *)queue restoreCompletedTransactionsFailedWithError:(NSError *)error
{
    PreyLogMessage(@"PreyStore Manager", 10, @"restoreCompletedTransactionsFailedWithError SKPaymentQueue");
   
    if (self.onRestoreFailed)
        self.onRestoreFailed(error);

    self.onRestoreFailed = nil;
}

- (void)paymentQueueRestoreCompletedTransactionsFinished:(SKPaymentQueue *)queue
{
    PreyLogMessage(@"PreyStore Manager", 10, @"paymentQueueRestoreCompletedTransactionsFinished SKPaymentQueue");

    if (self.onRestoreCompleted)
        self.onRestoreCompleted();
    
    self.onRestoreCompleted = nil;
}

- (void)paymentQueue:(SKPaymentQueue *)queue removedTransactions:(NSArray *)transactions
{
    PreyLogMessage(@"PreyStore Manager", 10, @"removedTransactions SKPaymentQueue");
}

#pragma mark Transaction Methods

- (void)completeTransaction:(SKPaymentTransaction *)transaction
{
    [self provideContent:transaction.payment.productIdentifier
              forReceipt:transaction.transactionReceipt
           hostedContent:nil];
    
    [[SKPaymentQueue defaultQueue] finishTransaction: transaction];
}

- (void)failedTransaction:(SKPaymentTransaction *)transaction
{
    NSLog(@"Failed transaction: %@", [transaction description]);
    NSLog(@"error: %@", transaction.error);
    
    [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
    
    if (self.onTransactionCancelled)
        self.onTransactionCancelled();
}

- (void)restoreTransaction:(SKPaymentTransaction *)transaction
{
    [self provideContent:transaction.originalTransaction.payment.productIdentifier
              forReceipt:transaction.transactionReceipt
           hostedContent:nil];
    
    [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
}

- (void)provideContent:(NSString*)productIdentifier forReceipt:(NSData*)receiptData hostedContent:(NSArray*)hostedContent
{
    PreyStoreProduct *thisProduct = [[PreyStoreProduct alloc] initWithProductId:productIdentifier receiptData:receiptData];
    
    [thisProduct verifyReceiptOnComplete:^
     {
         [self rememberPurchaseOfProduct:productIdentifier withReceipt:receiptData];
         
         if(self.onTransactionCompleted)
             self.onTransactionCompleted(productIdentifier, receiptData, hostedContent);
     }
                                     onError:^(NSError* error)
     {
         if (self.onTransactionCancelled)
             self.onTransactionCancelled(productIdentifier);
         else
             NSLog(@"The receipt could not be verified");
     }];
}

- (void)rememberPurchaseOfProduct:(NSString*)productIdentifier withReceipt:(NSData*)receiptData
{
    [PreyStoreManager setObject:[NSNumber numberWithBool:YES] forKey:productIdentifier];
    [PreyStoreManager setObject:receiptData forKey:[NSString stringWithFormat:@"%@-receipt", productIdentifier]];
}

#pragma mark Class Methods

+ (BOOL)isFeaturePurchased:(NSString*)featureId
{
    return [[PreyStoreManager numberForKey:featureId] boolValue];
}

+ (NSNumber*)numberForKey:(NSString*)key
{
    return [NSNumber numberWithInt:[[PreyStoreManager objectForKey:key] intValue]];
}

+ (id)objectForKey:(NSString*)key
{
    NSError *error = nil;
    id password = [SFHFKeychainUtils getPasswordForUsername:key andServiceName:@"PreyStore" error:&error];
    if(error) NSLog(@"%@", error);
    
    return password;
}

+ (void)setObject:(id)object forKey:(NSString*)key
{
    if (object)
    {
        NSString *objectString = nil;

        if ([object isKindOfClass:[NSData class]])
            objectString = [[NSString alloc] initWithData:object encoding:NSUTF8StringEncoding];
        
        if ([object isKindOfClass:[NSNumber class]])
            objectString = [(NSNumber*)object stringValue];
        
        NSError *error = nil;
        [SFHFKeychainUtils storeUsername:key andPassword:objectString forServiceName:@"PreyStore" updateExisting:YES error:&error];
        if(error) NSLog(@"%@", error);
    }
    else
    {
        NSError *error = nil;
        [SFHFKeychainUtils deleteItemForUsername:key andServiceName:@"PreyStore" error:&error];
        if(error) NSLog(@"%@", error);
    }
}

- (BOOL)removeAllKeychainData
{
    NSUInteger itemCount = self.purchasableObjects.count;
    NSError *error;
    
    //loop through all the saved keychain data and remove it
    for (int i = 0; i < itemCount; i++ ) {
        [SFHFKeychainUtils deleteItemForUsername:[self.purchasableObjects objectAtIndex:i] andServiceName:@"PreyStore" error:&error];
    }
    if (!error) {
        return YES;
    }
    else {
        return NO;
    }
}


@end

File: Prey/Classes/PreyStoreManager.swift
//
//  PreyStoreManager.swift
//  Prey
//
//  Created by Javier Cala Uribe on 14/07/16.
//  Copyright © 2016 Prey, Inc. All rights reserved.
//

import Foundation
import StoreKit

class PreyStoreManager: NSObject, SKProductsRequestDelegate, SKPaymentTransactionObserver {

    
    // MARK: Singleton
    
    static let sharedInstance   = PreyStoreManager()
    override fileprivate init() {
    }

    // MARK: Properties
    
    var purchasableObjects      = [SKProduct]()

    var productsRequest         :SKProductsRequest!
    
    var onTransactionCancelled  : [() -> Void] = []
    
    var onTransactionCompleted  : [(_ productId:String, _ receiptData:Data) -> Void] = []
    
    var onRestoreFailed         : [(_ error:Error) -> Void] = []
    
    var onRestoreCompleted      : [() -> Void] = []
    
    
    // MARK: Methods
    
    // Request Product Data
    func requestProductData() {
        
        var products = Set<String>()
        products.insert(subscription1Year)
    
        // Set productsRequest
        productsRequest             = SKProductsRequest(productIdentifiers:products)
        productsRequest.delegate    = self
        productsRequest.start()
        
        // Add Transaction Observer
        SKPaymentQueue.default().add(self)
    }
    
    // BuyFeature
    func buyFeature(_ featureId:String, onComplete:@escaping (_ productId:String, _ receiptData:Data) -> Void, onCancelled:@escaping () -> Void) {
        
        guard SKPaymentQueue.canMakePayments() else {
            displayErrorAlert("Check your parental control settings and try again later".localized,
                              titleMessage:"In-App Purchasing disabled".localized)
            return
        }

        onTransactionCompleted.append(onComplete)
        onTransactionCancelled.append(onCancelled)
        

        for product in purchasableObjects {
            if product.productIdentifier == featureId {
                let payment     = SKPayment(product:product)
                SKPaymentQueue.default().add(payment)
            }
        }
    }
    
    
    // MARK: Transaction Methods
    
    // CompleteTransaction
    func completeTransaction(_ transaction:SKPaymentTransaction) {

        guard let storeURL = Bundle.main.appStoreReceiptURL else {
            onTransactionCancelled.first?()
            return
        }
        
        guard let storeData = try? Data(contentsOf: storeURL) else {
            onTransactionCancelled.first?()
            return
        }
        
        provideContent(transaction.payment.productIdentifier, forReceipt:storeData)

        SKPaymentQueue.default().finishTransaction(transaction)
    }
    
    // FailedTransaction
    func failedTransaction(_ transaction:SKPaymentTransaction) {
        PreyLogger("Failed transaction")
        
        SKPaymentQueue.default().finishTransaction(transaction)

        onTransactionCancelled.first?()
    }
    
    // RestoreTransaction
    func restoreTransaction(_ transaction:SKPaymentTransaction) {
        
        guard let storeURL = Bundle.main.appStoreReceiptURL else {
            onTransactionCancelled.first?()
            return
        }
        
        guard let storeData = try? Data(contentsOf: storeURL) else {
            onTransactionCancelled.first?()
            return
        }

        guard let originalTransaction = transaction.original else {
            onTransactionCancelled.first?()
            return
        }
        
        provideContent(originalTransaction.payment.productIdentifier, forReceipt:storeData)

        SKPaymentQueue.default().finishTransaction(transaction)
    }
    
    // ProvideContent
    func provideContent(_ productIdentifier:String, forReceipt data:Data) {

        let product = PreyStoreProduct(initWithProductId:productIdentifier, receiptData:data)

        // Verify receipt
        product.verifyReceiptOnComplete({() in
            // Success
            self.savePurchaseOfProduct(productIdentifier, withReceipt:data)
            self.onTransactionCompleted.first?(productIdentifier, data)
            }, onError:{() in
            // Error
            self.onTransactionCancelled.first?()
        })
    }
    
    // Save purchase
    func savePurchaseOfProduct(_ productIdentifier:String, withReceipt data:Data) {
        UserDefaults.standard.set(true, forKey:productIdentifier)
    }
    
    // Check is feature purchased
    class func isFeaturePurchased(_ productIdentifier:String) -> Bool {
        return UserDefaults.standard.bool(forKey: productIdentifier)
    }
    
    
    // MARK: SKPaymentTransactinObserver
    
    // UpdteTransactions
    func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
        PreyLogger("updatedTransactions SKPaymentQueue")
        
        for transaction in transactions {
            switch transaction.transactionState {

            case .purchasing :
                PreyLogger("Purchasing")

            case .purchased :
                PreyLogger("Purchased")
                completeTransaction(transaction)

            case .failed :
                PreyLogger("Failed")
                failedTransaction(transaction)
            
            case .restored :
                PreyLogger("Restored")
                restoreTransaction(transaction)
            
            case .deferred :
                PreyLogger("Deferred")
                
            @unknown default:
                PreyLogger("Unknown")
            }
        }
    }
 
    // restoreCompletedTransactionsFailedWithError
    func paymentQueue(_ queue: SKPaymentQueue, restoreCompletedTransactionsFailedWithError error: Error) {
        PreyLogger("restoreCompletedTransactionsFailedWithError SKPaymentQueue")
        
        onRestoreFailed.first?(error)
        onRestoreFailed.removeAll()
    }
    
    // restoreCompletedTransactions
    func paymentQueueRestoreCompletedTransactionsFinished(_ queue: SKPaymentQueue) {
        PreyLogger("paymentQueueRestoreCompletedTransactionsFinished SKPaymentQueue")
        
        onRestoreCompleted.first?()
        onRestoreCompleted.removeAll()
    }
    
    // removedTransactions
    func paymentQueue(_ queue: SKPaymentQueue, removedTransactions transactions: [SKPaymentTransaction]) {
        PreyLogger("removedTransactions SKPaymentQueue")
    }
    
    // MARK: SKProductsRequest
    
    // DidReceiveResponse
    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
        PreyLogger("DidReceiveResponse SKProductsRequest")
        
        // Add object to purchableObjects from response
        purchasableObjects = response.products
        
        #if DEBUG
            for product in purchasableObjects {
                PreyLogger("Feature: \(product.localizedTitle), Cost: \(product.price.doubleValue)")
            }
        #endif
        
        // Reset productsRequest
        productsRequest = nil
    }
    
    // RequestDidFinish
    func requestDidFinish(_ request: SKRequest) {
        PreyLogger("RequestDidFinish SKProductsRequest")
    }
    
    // DidFailWithError
    func request(_ request: SKRequest, didFailWithError error: Error) {
        PreyLogger("DidFailWithError SKProductRequest: \(error.localizedDescription)")
        productsRequest = nil
    }
}
